design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+```
+where `TerminateQueryResponse` is again the same as `DdlResponse`/`DmlResponse`/`CreateSourceResponse`/`InsertIntoResponse`/`DropSourceResponse` above.
+
+The method name `terminatePersistenQuery(...)` is to distinguish from `terminatePushQuery(...)` below.
+
+#### Connectors
+
+```
+  CompletableFuture<ConnectorInfo> createSourceConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<ConnectorInfo> createSinkConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<Void> dropConnector(String name);
+```
+where `ConnectorInfo` is from an Apache Kafka module ([link](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/ConnectorInfo.java)).
+
+Or would we rather not have a dependency on Apache Kafka in the ksqlDB client interfaces?
+
+### Admin operations
+
+#### `SHOW TOPICS`
+
+```
+  CompletableFuture<List<TopicInfo>> listTopics();
+```
+with
+```
+public interface TopicInfo {
+
+  String getName();
+
+  int getPartitions();
+  
+  List<Integer> getReplicasPerPartition();
+
+}
+```
+
+#### `SHOW <STREAMS/TABLES>`
+
+```
+  CompletableFuture<List<StreamInfo>> listStreams();
+
+  CompletableFuture<List<TableInfo>> listTables();
+```
+with
+```
+public interface StreamInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+  
+}
+```
+and
+```
+public interface TableInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+
+  boolean isWindowed();
+  
+}
+```
+
+I'm not sure whether it makes more sense for `StreamInfo#getFormat()` and `TableInfo#getFormat()` to return a string or an enum value.
+The latter would make it easier for the user to know the possible values, but we'd have to keep the list up to date and would also sacrifice forward compatibility.
+
+#### `SHOW QUERIES`
+
+```
+  CompletableFuture<List<QueryInfo>> listQueries();
+```
+with
+```
+public interface QueryInfo {
+
+  boolean isPersistentQuery();
+
+  boolean isPushQuery();
+
+  /**
+   * Query ID, used for control operations such as terminating the query
+   */
+  String getId();
+
+  String getSql();
+
+  /**
+   * Name of sink, for a persistent query. Else, null.
+   */
+  String getSink();
+
+  /**
+   * Name of sink topic, for a persistent query. Else, null.
+   */
+  String getSinkTopic();
+  
+  /**
+   * Map of query state ("RUNNING", "ERROR", or "UNRESPONSIVE") to the number of ksqlDB servers
+   * on which this query is in each state.
+   */
+  Map<String, Integer> getServerStatusCounts();
+
+}
+```
+
+Again, not sure whether it makes more sense to represent query states as strings, or to introduce an Enum and have `getServerStatusCounts()` return an EnumMap instead.
+
+Does having `getSink()` and `getSinkTopic()` return null in the case of push queries make sense, or would we prefer an empty string or an Optional instead?
+
+#### `SHOW CONNECTORS`
```

Comment:
again, let's punt on connect support - I think we should let connect build a java client if that's what we want to do. Seems weird to have a client that talks to ksql, which just delegates to a connect cluster